server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -346,6 +374,67 @@ public String toString() {
         }
     }
 
+    @Override
+    protected TransportWriteAction.AsyncAfterWriteAction createAsyncAfterWriteAction(
+        IndexShard indexShard,
+        WriteRequest<?> request,
+        Translog.Location location,
+        TransportWriteAction.RespondingWriteResult respond,
+        Logger logger
+    ) {
+        return new AsyncAfterShardBulkAction(
+            indexShard,
+            request,
+            location,
+            respond,
+            logger,
+            latestSuccessfulLeaderCheckerTimeSupplier,
+            latestLeaderCheckThreshold
+        );
+    }
+
+    private static class AsyncAfterShardBulkAction extends AsyncAfterWriteAction {
+
+        private final IndexShard indexShard;
+        private final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier;
+        private final TimeValue leaderCheckerTimeThreshold;
+
+        public AsyncAfterShardBulkAction(
+            final IndexShard indexShard,
+            final WriteRequest<?> request,
+            @Nullable final Translog.Location location,
+            final RespondingWriteResult respond,
+            final Logger logger,
+            final LongSupplier latestSuccessfulLeaderCheckerTimeSupplier,
+            final TimeValue leaderCheckerTimeThreshold
+        ) {
+            super(indexShard, request, location, respond, logger);
+            this.indexShard = indexShard;
+            this.latestSuccessfulLeaderCheckerTimeSupplier = latestSuccessfulLeaderCheckerTimeSupplier;
+            this.leaderCheckerTimeThreshold = leaderCheckerTimeThreshold;
+        }
+
+        @Override
+        protected void finishOnNoPendingOps() {
+            if (indexShard.isRemoteTranslogEnabled() && elapsedTimeSinceLatestSuccessfulLeaderCheckBeforeThreshold()) {
+                respond.onFailure(
+                    new LeaderCheckerBeforeThresholdException(
```

Comment:
Why not cluster block exception since users are familiar with it. For some requests, users will see cluster block while for others LeaderCheckerBeforeThresholdException depending whether the check is done before or after. To me, this seems like an inconsistent behavior.

Reply:
This makes sense. Will incorporate this.